### ISSUE TITLE
Reintegrate the RH-certified fork back into the main line

### ISSUE THREAD

**Issue by mflendrich** [Kong/kong-operator#24]
https://github.com/Kong/kong-operator/tree/fork/certified is a fork of kong-operator which makes kong-operator RH certified.
Maintaining two independent codebases long-term does not look like a viable option because of:
- the need to reintegrate changes both ways
- the risk of code drift
- double the testing effort

The scope of this issue is:
- to make it possible to deliver both a "standard" and a "RH-certified" operator from the same git commit
- builder/contributor/usage documentation

Two ideas for implementation:
- recipes for two "flavors" of the kong-operator image
- (maybe not possible) one "flavor" of the kong-operator image, but with the default configuration chosen by a runtime parameter

**Comment by mflendrich** [comment]
I recommend doing #20 first so that we first have some regression test in place.

**Comment by rainest** [comment]
We probably need to maintain the fork at some level, but we can probably make it lighter-weight: as-is, the fork branch has both some necessary permanent differences (using different images, e.g. https://github.com/Kong/kong-operator/commit/06ec06dbbfb0f16187b1e6b34bcc5e099a65991a and https://github.com/Kong/kong-operator/commit/ffe2537cbcfacf07e1db8ab0be68ca4c1c3f0824) and some things that should be re-integrated (a bunch of the other commits that I made to past RH verification and forgot what was or wasn't specific to the certified operator).

So far as I know, we do wish to continue releases that pull images from our Bintray registries and use images that aren't RHEL-based, and wish to use those as the default for our non-certified operator releases. That and RH's requirements are opposite and impossible to reconcile, so bare minimum we need to use the override mechanism to accomplish that. However, the cost of maintaining that fork is pretty minimal, since it's changing like 3 lines to swap out one image URI for another.

There are some things we *may* be able to combine, e.g. I know RH wants us to use different [package names](https://github.com/Kong/kong-operator/commit/c58b868a39b838b6d9731d8e811aaa883ebfeb8f#diff-70a688227205df4ed6c1c883ccd86ac2R4) for the certified and not certified versions. As-is I don't know of a way to reconcile it, but maybe it's possible to do so via channels or something.

On the chart end, we currently use a BusyBox init container to block on Postgres availability, and cannot do so for the certified operator because BusyBox ain't RHEL-based, and no UBI offers the same tools out of the box. While we could produce an RHEL-based image that does include netcat, we don't really want to because it's an extra artifact we have to maintain. Instead, the certified operator replaces the BusyBox container with a UBI container that then runs some [fake netcat bash script](https://github.com/Kong/charts/commit/86b306b92b828df8b9dfd999481173aa0d1deb8e). We can probably use that same script with whatever image (it should run, for instance, in the base alpine and centos images), but switching over has the potential to cause some unexpected breakage (we don't know of any from our own testing, but don't have user experiences that corroborate that yet).

**Comment by mflendrich** [comment]
Assuming that we need to maintain two different parallel releases of the operator - one "vanilla" and one "RH-certified", two thoughts here:

1. We can have a separate Dockerfile for the "vanilla" operator and for the RH-certified operator. Those two Dockerfiles would live side by side. Our CI would test both flavors of the operator in parallel.
2. We'd need two sets of `deploy/` manifests. Kustomize (or a similar k8s config de-dup tool) could help to reduce verbosity.
3. For diverging Helm charts: the RH-certified operator could use a thin wrapper chart over the vanilla chart, only changing the essential defaults
4. For `olm/`: it seems that OperatorHub and RHM rely on completely separate release procedures:

    - OperatorHub asks you to make a PR to their GitHub repo
    - RHM asks you to upload a Docker image for scanning, and upload a .zip with metadata

    RHM and OperatorHub release metadata have a similar format but different content (RHM requires extra settings and a separate name) so it seems justified to keep them in the repo in parallel.

If we choose that reintegration is worth the burden, does this sound like a reasonable plan?

**Comment by rainest** [comment]
tl;dr probably we should just decide if there's any reason to keep the chart split: probably there isn't, and if not we don't need multiple Dockerfiles. If Kustomize can handle the deployment.yaml and OLM CRD patches, we can probably use that to start checking in parallel deploy/OLM file trees. As the actual release artifacts aren't _closely_ tied to this repo that's not as large a concern as I thought, though we'll eventually want some automation to change the OLM directory names (not sure if operator-sdk lets you specify one when building) for builds.

1. The Dockerfiles are currently identical (and both out of date--whoops, that `LABEL` is pretty old) since it just says "start from the helm-operator image and copy in our chart/waches.yaml). The chart contents are different, but the Dockerfile isn't aware of that, it just knows what's on the filesystem when it's built. If we do handle both in mainline, then agreed, that does change, but it's pretty minor (one path is different).

2. Kustomize should work fine here. AFAIK only the image in `deploy/operator/deployment.yaml` should actually change, which Kustomize should be more than capable of. There are other divergences if you actually diff the branches, but those all appear to be missed updates that should be in sync with both.

3. Not doable as proposed. The chart differences involve template changes, and values.yaml overrides alone cannot handle those. I think that's better handled by removing the existing standard busybox image/template for wait-for-db and using the bash hack version for both: I don't recall it introducing any significant change in behavior, so it should be fine for either. I don't love the bash hack option, but it works.

4. Will Kustomize work here also? The stuff that needs to change for OLM is a CRD, so it should, I think. I wouldn't necessarily worry about the different publish mechanisms quite yet, as if we can make the metadata/package content changes in parallel, the rest is simpler. Something like:

  - Perform general updates as needed, build standard images, cut a release, and push images.
  - Copy OH.io-relevant metadata into our fork of their repo, PR.
  - Apply Kustomize changes for the certified operator somewhere (temporary local branch or similar), build image, rearrange OLM files into their flat format, zip up, and send both the image and bundle to Red Hat.

A Makefile to cut out some of the toil around the final step would be nice, but last go round I had more pain from making sure I got the changes right. If Kustomize can perform those, the remaining parts (run an image build, copy files into the flat structure, push things) aren't too much of a burden.

For OLM in particular, part of the difficulty there was needing to apply those changes across many files, but I think that could probably be handled pretty well--assuming the CRD structure isn't for what Kustomize needs to change isn't likely to change often, maybe make a parallel certified-olm that we check in and update for each release using the Kustomize patch, and update the patch in the rare event it needs a change.

**Comment by rainest** [comment]
One thing to note is that the linked chart PR removes this disclaimer, since it's Red Hat-specific:

```
# The PostgreSQL chart is not compatible with Red Hat's official Postgres
# images. If you use the operator in an offline environment, this must be
# disabled, and you must configure your environment variables to use an
# independently-installed database or to use DB-less mode.
```
We should find a way to work that back into the certified operator's example Kong CRD.

**Comment by mflendrich** [comment]
timebox: 1.5 days

**Comment by mflendrich** [comment]
Won't fix because we're doing golang operator. We'll live with the forks for now. Also, #70 is a duplicate of this.
